Libs/DICOM/Core/ctkDICOMQueryJob.h

```diff
@@ -81,9 +81,7 @@ class CTK_DICOM_CORE_EXPORT ctkDICOMQueryJob : public ctkDICOMJob
   ///@{
   /// Server
   Q_INVOKABLE ctkDICOMServer* server() const;
```

Comment:
Does this have to be public? If it is only for Python then we could add an #ifdef around it to hide from C++.

Reply:
yes I need it pubblic in C++. We call that method in the worker (the worker need access to the server pointer saved in the job object)

Reply:
If you need a shared pointer then it means that the ownership of the object is complex. It is better to always use shared pointer for accessing the object and not mix plain and shared pointer.

Reply:
you are right. Before the server pointer were all shared pointer owned by the scheduler. However I have experienced crash, because you can easily crash if you do any modification in the mean thread on the server shared pointers and meanwhile a job is refering to it in another thread (e.g. if you delete a server but there are still jobs running). But I agree also with your comment. I change the job classes to own a copy of the server, which will be simple pointer and not a QSharedPointer.

Reply:
Summary of what I have applied:
1) to ensure thread safety `ctkDICOMRetrieveJob` and `ctkDICOMQueryJob` the will own a copy of the server pointer, i.e.:

```
void ctkDICOMRetrieveJob::setServer(const ctkDICOMServer& server)
{
  Q_D(ctkDICOMRetrieveJob);
  d->Server = server.clone();
}

```

where d->Server is a plain pointer.

2) the main copy of the server objects are still owned by the `ctkDICOMScheduler` as a `QSharedPointer<ctkDICOMServer>` (i.e. `QList<QSharedPointer<ctkDICOMServer>> Servers;`)

3) we could make all the methods relative to the server thread safe with a Mutex, but I really think it would just make the full infrastructure slower and very complicated. Few copies around of a server object should be reasonable (the server object attributes is just strings and intergers).